File: scripts/geocode-addresses.js
const fs = require('fs');
const path = require('path');
const https = require('https');

const API_KEY = 'AIzaSyDt7ieN0wG23Zy5ZCuxg0pjHNqowquaZHI';
const affiliatesPath = path.join(__dirname, '../data/affiliates.json');
const docsPath = path.join(__dirname, '../docs/data/affiliates.json');
const affiliates = JSON.parse(fs.readFileSync(affiliatesPath, 'utf8'));

function httpsGet(url) {
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => {
        try { resolve(JSON.parse(data)); } catch (e) { reject(e); }
      });
    }).on('error', reject);
  });
}

async function geocodeAddress(address) {
  const url = `https://maps.googleapis.com/maps/api/geocode/json?address=${encodeURIComponent(address)}&key=${API_KEY}&region=kr`;
  try {
    const data = await httpsGet(url);
    if (data.status === 'OK' && data.results.length > 0) {
      const location = data.results[0].geometry.location;
      return { latitude: location.lat, longitude: location.lng };
    }
    return null;
  } catch (error) {
    console.log(`  ❌ 오류: ${error.message}`);
    return null;
  }
}

async function findPlace(name, address) {
  const query = `${name} ${address}`;
  const url = `https://maps.googleapis.com/maps/api/place/findplacefromtext/json?input=${encodeURIComponent(query)}&inputtype=textquery&fields=place_id,photos&key=${API_KEY}&region=kr`;
  try {
    const data = await httpsGet(url);
    if (data.status === 'OK' && data.candidates.length > 0) {
      const place = data.candidates[0];
      if (place.photos && place.photos.length > 0) {
        return `https://maps.googleapis.com/maps/api/place/photo?maxwidth=800&photo_reference=${place.photos[0].photo_reference}&key=${API_KEY}`;
      }
    }
    return 'https://placehold.co/200x200';
  } catch (error) {
    return 'https://placehold.co/200x200';
  }
}

function delay(ms) { return new Promise(resolve => setTimeout(resolve, ms)); }

async function main() {
  console.log('🔍 주소가 있는 업체 중 좌표가 기본값인 업체를 찾습니다...\n');
  let updatedCount = 0;
  const defaultLat = 33.4996;
  const defaultLng = 126.5312;

  for (let i = 0; i < affiliates.length; i++) {
    const affiliate = affiliates[i];
    if (affiliate.address && affiliate.address.trim() !== '' && 
        affiliate.latitude === defaultLat && affiliate.longitude === defaultLng) {
      console.log(`\n[${updatedCount + 1}] ${affiliate.name}`);
      console.log(`  주소: ${affiliate.address}`);
      
      const geoResult = await geocodeAddress(affiliate.address);
      if (geoResult) {
        affiliates[i].latitude = geoResult.latitude;
        affiliates[i].longitude = geoResult.longitude;
        console.log(`  ✓ 좌표: ${geoResult.latitude}, ${geoResult.longitude}`);
        
        const imageUrl = await findPlace(affiliate.name, affiliate.address);
        affiliates[i].image = imageUrl;
        console.log(`  ✓ 이미지: ${imageUrl.substring(0, 50)}...`);
        
        updatedCount++;
        await delay(200);
      }
    }
  }

  if (updatedCount > 0) {
    const jsonData = JSON.stringify(affiliates, null, 2);
    fs.writeFileSync(affiliatesPath, jsonData, 'utf8');
    console.log('✓ /data/affiliates.json 저장됨');

    fs.writeFileSync(docsPath, jsonData, 'utf8');
    console.log('✓ docs/data/affiliates.json 복사됨');

    console.log(`\n✅ 완료! ${updatedCount}개 업체 업데이트됨`);
  } else {
    console.log('\n✅ 업데이트할 업체가 없습니다.');
  }
}

main().catch(console.error);

File: scripts/geocode_addresses.py
import json
import time
import urllib.parse
import urllib.request
from pathlib import Path

API_KEY = "AIzaSyDt7ieN0wG23Zy5ZCuxg0pjHNqowquaZHI"
DEFAULT_LAT = 33.4996
DEFAULT_LNG = 126.5312
DELAY_SEC = 0.2

ROOT = Path(__file__).resolve().parents[1]
AFFILIATES_PATH = ROOT / "data" / "affiliates.json"
DOCS_PATH = ROOT / "docs" / "data" / "affiliates.json"


def fetch_json(url: str):
    request = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(request, timeout=30) as response:
        data = response.read().decode("utf-8")
        return json.loads(data)


def geocode(address: str):
    params = {
        "address": address,
        "key": API_KEY,
        "region": "kr",
    }
    url = "https://maps.googleapis.com/maps/api/geocode/json?" + urllib.parse.urlencode(params)
    data = fetch_json(url)
    if data.get("status") == "OK" and data.get("results"):
        location = data["results"][0]["geometry"]["location"]
        return location["lat"], location["lng"]
    return None


def find_place_photo(name: str, address: str):
    query = f"{name} {address}".strip()
    params = {
        "input": query,
        "inputtype": "textquery",
        "fields": "place_id,photos",
        "key": API_KEY,
        "region": "kr",
    }
    url = "https://maps.googleapis.com/maps/api/place/findplacefromtext/json?" + urllib.parse.urlencode(params)
    data = fetch_json(url)
    if data.get("status") == "OK" and data.get("candidates"):
        photos = data["candidates"][0].get("photos")
        if photos:
            photo_ref = photos[0].get("photo_reference")
            if photo_ref:
                params = {
                    "maxwidth": 800,
                    "photo_reference": photo_ref,
                    "key": API_KEY,
                }
                return "https://maps.googleapis.com/maps/api/place/photo?" + urllib.parse.urlencode(params)
    return None


def main():
    affiliates = json.loads(AFFILIATES_PATH.read_text(encoding="utf-8"))
    to_update = [a for a in affiliates if a.get("address") and a.get("latitude") == DEFAULT_LAT and a.get("longitude") == DEFAULT_LNG]

    print(f"총 {len(affiliates)}개 중 {len(to_update)}개 업체를 업데이트합니다.\n")
    updated = 0
    failures = []

    for idx, affiliate in enumerate(to_update, 1):
        name = affiliate.get("name")
        address = affiliate.get("address")
        print(f"[{idx}/{len(to_update)}] {name}")
        print(f"  주소: {address}")

        try:
            geo = geocode(address)
        except Exception as exc:
            print(f"  ❌ 지오코딩 오류: {exc}")
            failures.append(name)
            time.sleep(DELAY_SEC)
            continue

        if not geo:
            print("  ❌ 좌표를 찾지 못했습니다.")
            failures.append(name)
            time.sleep(DELAY_SEC)
            continue

        lat, lng = geo
        affiliate["latitude"] = lat
        affiliate["longitude"] = lng
        print(f"  ✓ 좌표: {lat}, {lng}")

        photo_url = None
        try:
            photo_url = find_place_photo(name or "", address)
        except Exception as exc:
            print(f"  ⚠️  이미지 조회 실패: {exc}")

        if photo_url:
            affiliate["image"] = photo_url
            print("  ✓ 이미지 갱신")
        else:
            print("  ⚠️  이미지를 찾지 못해 기본 이미지를 유지합니다.")

        updated += 1
        time.sleep(DELAY_SEC)

    if updated:
        json_data = json.dumps(affiliates, ensure_ascii=False, indent=2)
        AFFILIATES_PATH.write_text(json_data, encoding="utf-8")
        DOCS_PATH.write_text(json_data, encoding="utf-8")
        print(f"\n✅ 완료! {updated}개 업체 업데이트됨")
    else:
        print("\n정보를 업데이트할 업체가 없습니다.")

    if failures:
        print("\n다음 업체는 좌표 또는 이미지를 찾지 못했습니다:")
        for name in failures:
            print(f" - {name}")


if __name__ == "__main__":
    main()
